Import geant4maker.

klendathu2k opened the pull request:
Draft PR for geant4star.  Supersedes #617.  

Branch was rebased against current main (02/21/24).    Plus few additional fixes to get it to run (up to loading G3) in the current environment.

NOTE:  The code should be functional, but requires two modules which we don't currently have in spack, plus an update to  a newer version of the geant4 library.  Specifically, we need to add a GEANT3 VMC library, and update the Geant 4 VMC library to one which supports multi-engine simulations.

- [x] [GEANT3 VMC 3.9p1](https://github.com/vmc-project/geant3/releases/tag/v3-9-p1)
- [x] [Geant4 10.5.1](https://github.com/Geant4/geant4/releases/tag/v10.5.1)
- [x] [Geant 4 VMC 5.0](https://github.com/vmc-project/geant4_vmc/releases/tag/v5-0)

These versions have been tested together by the VMC team. 

I should also mention that we are at a point of divergence in the ROOT / VMC development.  Currently we are using the VMC version 1.1, which is built with ROOT.  The next version is 2.0, which is shipped separately.  The GEANT3 VMC 3.9p1 release includes fixes to sensitive volumes which we require, and is the last version able to run with VMC 1.1.  Any move to a new version of ROOT, VMC and/or Geant (either G3 or G4) likely entails advancing to a new version of all of these packages... with the _possible_ exception of ROOT.

![geant4star](https://private-user-images.githubusercontent.com/56083924/279710938-ee254949-dcb8-4c4d-9d0d-347e5f55d834.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg1MzgwNTMsIm5iZiI6MTcwODUzNzc1MywicGF0aCI6Ii81NjA4MzkyNC8yNzk3MTA5MzgtZWUyNTQ5NDktZGNiOC00YzRkLTlkMGQtMzQ3ZTVmNTVkODM0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIxVDE3NDkxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM2NjBjMWY2NWQ0NmMyMmNkODY0MGEwZTU0MjFkZDAwNjQ1MGE5ZjIxMjQ0YjEyYTUxNTgxZjBjYTMxNDk4OTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.otPtKaE8iMbKsLiKA50bMFqaYXF2UN8_UpeL2xzxLfA)

Code compiles against root 6.24 in our production environment:

```
starver dev config/v0.3.0-rhel7-root6.24.06
setenv STAR .
cons
```

- [x] StGeant4Maker should be excluded from all ROOT5 builds (in CI)
- [x] StGeant4Maker should be excluded from 32 bit builds
- [x] root5 and root6 builds should pass CI tests

klendathu2k committed:
[g4star] Expose partcile stack for the geant4 application.

[g4star] Update stacker and agml extension classes to enable multi engine tracking and user defined hits in the geant4 application.  Few additional fixes and cleanups in the code.

[g4star] Import the StGeant4Maker into the repository.  Builds under root 6.24 (with some additional compilation flags).

[g4star] Defer mysql load until StBFChain can do it.

[g4star] Add geant4vmc, geant4mk and fastjet chain options.  Load libfastjet with stargen package b/c one or more filters depends on it.  (Should consider splitting filters into sep chain opt).

[g4star] Cleanup commented out includes.

[g4star] Reduce compile time warnings...

[g4star] ... more reduction of compiler warnings

... and even more reduction of compiler warnings...

[g4star] And modify cons to pickup the include paths to geant4 and geant4 vmc libraries

klendathu2k committed:
[g4star] Expose partcile stack for the geant4 application.

[g4star] Update stacker and agml extension classes to enable multi engine tracking and user defined hits in the geant4 application.  Few additional fixes and cleanups in the code.

[g4star] Import the StGeant4Maker into the repository.  Builds under root 6.24 (with some additional compilation flags).

[g4star] Defer mysql load until StBFChain can do it.

[g4star] Add geant4vmc, geant4mk and fastjet chain options.  Load libfastjet with stargen package b/c one or more filters depends on it.  (Should consider splitting filters into sep chain opt).

[g4star] Cleanup commented out includes.

[g4star] Reduce compile time warnings...

[g4star] ... more reduction of compiler warnings

... and even more reduction of compiler warnings...

[g4star] And modify cons to pickup the include paths to geant4 and geant4 vmc libraries

klendathu2k committed:
[geant4star] Load packages required for G4* for 64bit builds.

klendathu2k committed:
[geant4star] Load packages required for G4* for 64bit builds.

klendathu2k committed:
... and the CLHEP package required by G4 as well ...

klendathu2k committed:
... and the CLHEP package required by G4 as well ...

klendathu2k committed:
Fix duplicate.

klendathu2k committed:
Fix duplicate.

klendathu2k committed:
Restore gcomad for agstar dummy routines

klendathu2k committed:
Restore gcomad for agstar dummy routines

klendathu2k committed:
Extern causes issues at runtime

klendathu2k committed:
Extern causes issues at runtime

klendathu2k committed:
Duplicate load removed.

klendathu2k committed:
Duplicate load removed.

klendathu2k committed:
Make multi-engine simulation the default.  Comment out the "NAME:engine" option (which was meant as documentation).

klendathu2k committed:
Make multi-engine simulation the default.  Comment out the "NAME:engine" option (which was meant as documentation).

klendathu2k commented on StRoot/StGeant4Maker/AgMLBTofVolumeId.h:
AgMLBTofVolumeId and similarly named classes implement the unique volume IDs defined in pams/sim/g2t/g2t_volume_id.g.  They provide the unique associated between hits and hardware channels.

klendathu2k commented on StRoot/StGeant4Maker/AgStarDummy.cxx:
AgStarDummy inserts a few dummy routines (mangled to look like FORtran routines) which satisfy dependencies between the StarGenerator package and starsim.

klendathu2k commented on StRoot/StarGenerator/BASE/StarPrimaryMaker.h:
```diff
@@ -167,6 +167,8 @@ class StarPrimaryMaker : public StMaker
 
   /// Return a pointer to the event
   StarGenEvent *event() { return mPrimaryEvent; }
+  /// Return a pointer to the stack 
```
Exposes the particle stack to StGeant4Maker for PushPrimaries.

klendathu2k commented on StarVMC/StarAgmlLib/AgMLExtension.cxx:
```diff
@@ -9,9 +10,31 @@ AgMLExtension::AgMLExtension() : TGeoRCExtension(),
+
+}
+
+std::map<TString, AgMLExtension*> AgMLExtension::mExtensionMap;
```
The GEANT3 VMC took use of both the "user" and "framework" extensions to TGeoVolume, leaving us with the only option of keeping a map between volume names and extension data.

iraklic reviewed (approved):
does this mean we did not have btof and eemc in AgML format before? (I though we did)

klendathu2k commented:
> does this mean we did not have btof and eemc in AgML format before? (I though we did)

I assume you are referring to the "volume ID" classes?   No.  These are new classes to support the geant4star application.

klendathu2k commented on mgr/Conscript-standard:
```diff
@@ -720,6 +720,22 @@ if ( $pkg !~ /^sim$/ && $pkg !~ /^gen$/ ) {
 
+    } elsif ( $pkg =~ m/StGeant4Maker$/ ) {
+
+	# Add G4 to include path
```
Note... we may need to also add a geant3 vmc path to the build.  TBD.

plexoos commented:
If you have a chance please try to build with this dependencies

```
module load geant4-vmc-5-0-p5-root-6.24.06 geant4-10.5.1 geant3-3-8-root-6.24.06
```

Geant3-3-9-p1 is currently missing from the spack repo so, I'll need to add it

klendathu2k commented:
Hi Dmitri,

Will do.  The changes in Geant3-3-9-p1 are w/ respect to sensitive 
detector functionality,
not interface, so building against it will at least allow me to test a 
pure G4 simulation.

Thanks,
Jason

On 2024-03-06 09:28, Dmitri Smirnov wrote:
> If you have a chance please try to build with this dependencies
> 
> module load geant4-vmc-5-0-p5-root-6.24.06 geant4-10.5.1
> geant3-3-8-root-6.24.06
> 
> Geant3-3-9-p1 is currently missing from the spack repo so, I'll need
> to add it
> 
> --
> Reply to this email directly, view it on GitHub [1], or unsubscribe
> [2].
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
> 
> Links:
> ------
> [1] 
> https://github.com/star-bnl/star-sw/pull/659#issuecomment-1980993430
> [2] 
> https://github.com/notifications/unsubscribe-auth/ANL4LVED3RAU7JXGIH4UKPDYW4RXBAVCNFSM6AAAAABDTP5ZLOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOBQHE4TGNBTGA

plexoos commented:
geant3@3-9-p1 is also available on the SDCC machines now. So, the following modules can be used for testing:

```
module load geant4-vmc-5-0-p5-root-6.24.06 geant4-10.5.1 geant3-3-9-p1-root-6.24.06
```

klendathu2k committed:
Handle ROOT version dependence for VMC library.

klendathu2k committed:
Handle ROOT version dependence for VMC library.

klendathu2k committed:
Add VMC include path if separate from ROOT.

klendathu2k committed:
Add VMC include path if separate from ROOT.

klendathu2k committed:
Load VMC, G3/G4 VMC and concrete libraries for 64bit builds.

klendathu2k committed:
Load VMC, G3/G4 VMC and concrete libraries for 64bit builds.

klendathu2k commented:
With the last three pushes, code compiles and runs the TPC to completion.  Albeit with no hits generated.  Reason TBC.

genevb commented on StRoot/StBFChain/BigFullChain.h:
```diff
@@ -1294,8 +1302,8 @@ Bfc_st BFC[] = { // standard chains
-  {"geant4vmc",   "","", "geant4", "",                    "libVMC.so,libgeant4vmc.so", "Load G4 VMC libs", false},
-  {"geant4mk",    "","", "stargen,geant4vmc",  "", "StGeant4Maker.so,StarMagField.so", "Load G4 VMC libs", false},
+  {"geant4vmc",   "","", "geant4", "",                    libVMC ",libgeant4vmc.so", "Load G4 VMC libs", false},
+  {"geant4mk",    "","", "stargen,geant4vmc,geant3vmc",  "", "StGeant4Maker.so,StarMagField.so", "Load G4 VMC libs", false},
```
Geant4 needs Geant3? Just making sure this wasn't a mistake.

klendathu2k replied:
The geant4maker is a multi-physics-engine application.  We will, for example, be able to run G3 at mid rapidity and G4 in the forward direction.  Or run G3 in the TPC and G4 in the calorimeters.  Or ...

klendathu2k commented:
Short answer is no.  G4 does not need G3 to compile or run.

Long answer...  the G3 to G4 library provides support for exporting G3 
geometry objects
(materials, media, volumes, ...) to G4 geometry objects.  Also ROOT 
geometries.  VMC
applications that use ROOT geometries for track navigation still need to 
export their
material and tracking media to G4 for  physics simulation.


On 2024-03-12 20:37, Gene Van Buren wrote:
> @genevb commented on this pull request.
> -------------------------
> 
> In StRoot/StBFChain/BigFullChain.h [1]:
> 
>> @@ -1294,8 +1302,8 @@ Bfc_st BFC[] = { // standard chains
>    {"g4geant3",    "", "", "",                     "", "libG3toG4.so",
> "Load g3 to g4 support", false },
>    {"geant4",      "", "", "g4physics,g4interfaces","","","Load G4
> libs", false},
> 
> -  {"geant4vmc",   "","", "geant4", "",
> "libVMC.so,libgeant4vmc.so", "Load G4 VMC libs", false},
> -  {"geant4mk",    "","", "stargen,geant4vmc",  "",
> "StGeant4Maker.so,StarMagField.so", "Load G4 VMC libs", false},
> +  {"geant4vmc",   "","", "geant4", "",                    libVMC
> ",libgeant4vmc.so", "Load G4 VMC libs", false},
> +  {"geant4mk",    "","", "stargen,geant4vmc,geant3vmc",  "",
> "StGeant4Maker.so,StarMagField.so", "Load G4 VMC libs", false},
> 
> Geant4 needs Geant3? Just making sure this wasn't a mistake.
> 
> --
> Reply to this email directly, view it on GitHub [2], or unsubscribe
> [3].
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
> 
> Links:
> ------
> [1] https://github.com/star-bnl/star-sw/pull/659#discussion_r1522311684
> [2] 
> https://github.com/star-bnl/star-sw/pull/659#pullrequestreview-1932975768
> [3] 
> https://github.com/notifications/unsubscribe-auth/ANL4LVAPKQEVTJBUK5JYYITYX6NUFAVCNFSM6AAAAABDTP5ZLOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSMZSHE3TKNZWHA

klendathu2k committed:
Invoke local copy for sensitive volume materials.

klendathu2k committed:
Invoke local copy for sensitive volume materials.

klendathu2k committed:
Different set of sensitive volumes in sTGC geometry.

klendathu2k committed:
Different set of sensitive volumes in sTGC geometry.

klendathu2k committed:
Think we really want muons here...

klendathu2k committed:
Think we really want muons here...

klendathu2k committed:
Volume sensitivitiy can be flagged in either the material or medium blocks.

klendathu2k committed:
Volume sensitivitiy can be flagged in either the material or medium blocks.

klendathu2k committed:
... and definitely want them here...

klendathu2k committed:
... and definitely want them here...

klendathu2k committed:
libgeom not needed for geant3vmc / fix EEmcUtil chain option.

klendathu2k committed:
libgeom not needed for geant3vmc / fix EEmcUtil chain option.

klendathu2k committed:
Add g3 vmc as default chain option.

klendathu2k committed:
Add g3 vmc as default chain option.

klendathu2k committed:
libPhysics does not need to be preloaded.

klendathu2k committed:
libPhysics does not need to be preloaded.

klendathu2k committed:
Reduce unneccesary verbosity.

klendathu2k committed:
Reduce unneccesary verbosity.

klendathu2k committed:
Apply the NODEBUG environment variable.

klendathu2k committed:
Apply the NODEBUG environment variable.

klendathu2k commented on StRoot/StGeant4Maker/StGeant4Maker.cxx:
```diff
@@ -0,0 +1,1539 @@
+#include <cstring>
+//________________________________________________________________________________________________
+
+// Functors used to copy the hits from the sensitive detector hit collections into the g2t tables.
```
These functors are the equivalent of the `pams/sim/g2t/g2t_???.F` functions, which copy the zebra hits into the c++ tables.

klendathu2k commented on StRoot/StGeant4Maker/StGeant4Maker.cxx:
```diff
@@ -0,0 +1,1539 @@
+TGeant3TGeo* gG3 = 0;
+
+// Function to process one event
+std::function<void(void)> trigger;
```
The function which processes one event is different depending on whether we are running a pure G3, pure G4 or mixed simulation.  So we assign this to a functional placed in the global scope of the module.   The trigger function is created on line 609.

klendathu2k commented on StRoot/StGeant4Maker/StGeant4Maker.cxx:
```diff
@@ -0,0 +1,1539 @@
+//________________________________________________________________________________________________
+StarParticleData &particleData = StarParticleData::instance();
+//________________________________________________________________________________________________
+StarVMCApplication::StarVMCApplication( const Char_t *name, const Char_t *title, double zmax, double rmax, std::string multi, StMCParticleStack* stack ) : 
```
You will note that there are two class implementations in this file.  StGeant4Maker and StarVMCApplication.  The VMC application creates and steers the simulation, while the StGeant4Maker serves as the interface and bridge between our framework and the application.

plexoos commented on StRoot/StBFChain/doc/StBFChain.cxx_doc:
This auto generated file should not be modified

plexoos commented on StRoot/StBFChain/doc/index.html:
This auto generated file should not be modified

plexoos commented on StRoot/StBFChain/BigFullChain.h:
```diff
@@ -25,6 +25,14 @@ ________________________________________________________________________________
+#if ROOT_VERSION_CODE>=ROOT_VERSION(6,24,0)
+#define libVMC "libVMCLibrary.so"
+#else
+#define libVMC "libVMC.so"
```
Should use all caps for the macro name

klendathu2k replied:
Done

klendathu2k committed:
Do not build geant4star if we are not at 64bit and we are not at least ROOT version 6.24.

klendathu2k committed:
Remove dead code.

klendathu2k committed:
Cleanup comments.

klendathu2k committed:
Merge branch 'import-geant4maker-rebased' of https://github.com/klendathu2k/star-sw-1 into import-geant4maker-rebased

klendathu2k committed:
libVMC --> LIBVMC

klendathu2k committed:
Problem between chair and keyboard...  the EEmcUtil class is provided in the StEEmcUtil library.  SMH.

klendathu2k committed:
ROOT5 / CINT doesn't replace the LIBVMC definition which concatinating strings for some reason.  It also does not play nice with the ROOT_VERSION macro.  So we do this the ugly way.

klendathu2k committed:
Merge branch 'main' into import-geant4maker-rebased

klendathu2k committed:
Prevent ROOT6/CLING from loading all TROOT exposed classes twice... (which couldn't possibly cause any side effects, but lets be safe about it shall we?)

klendathu2k committed:
Local code implies local search path for headers.

klendathu2k committed:
Cling wrap... to prevent ROOT from ingesting class definitions multiple times.

...

klendathu2k committed:
From PR#693

...

klendathu2k committed:
Merge branch 'star-bnl:main' into main

klendathu2k committed:
When geant4maker is set, we should add the geantBranch to the output tree.

klendathu2k committed:
Add pythia6 chain option.

klendathu2k committed:
Output implies geant4out, not geantout.

klendathu2k committed:
Option to load geant tables w/out other dependencies.

klendathu2k committed:
The protected against both CINT and CLING.

klendathu2k committed:
Remove ancient version of geant3 vmc from our source tree.

klendathu2k committed:
TGeant3TGeo available under the TGeant3/ directory.

klendathu2k committed:
...

klendathu2k committed:
Use R__LOAD_LIBRARY rather than initialization macro.  Add valgrind options to the geant4star script.

klendathu2k committed:
This removes the internal CLHEP library distributed with G4, so we can use the external CLHEP library via spack.  NOTE... this may not be correct / consistent for G4.  We may need to revisit.

klendathu2k committed:
Use R__LOAD_LIBRARY ...

klendathu2k committed:
Clingwrap header files.

klendathu2k committed:
Add (possibly nonfunctional) hijing macro.

klendathu2k committed:
Will need to expose TCallf77 in 32bit builds as well.  So make sure we have the geant3 library available.

klendathu2k committed:
Merge branch 'star-bnl:main' into main

klendathu2k committed:
Handle the (hopefully rare) case where truth track is not defined.

klendathu2k committed:
Expose enum to cling.

klendathu2k committed:
Merge branch 'main' into import-geant4maker-rebased

klendathu2k committed:
Check for null pointers and skip to next.

klendathu2k committed:
Merge branch 'import-geant4maker-rebased' of https://github.com/klendathu2k/star-sw-1 into import-geant4maker-rebased

klendathu2k committed:
No need to update for ROOT 5.

klendathu2k committed:
Merge branch 'star-bnl:main' into main

klendathu2k committed:
Merge branch 'import-geant4maker-rebased' into main-patchup

klendathu2k committed:
Merge pull request #21 from klendathu2k/main-patchup

Merge upstream changes from main into the geant4maker integration branch.

klendathu2k committed:
Uncomment the line which enables geometry selection.

klendathu2k committed:
Readout FMS hits.

klendathu2k committed:
Silently skip nonsensitive volumes.

klendathu2k committed:
The FMS is a calorimeter.

klendathu2k committed:
Silence is golden.

klendathu2k committed:
FMS sensitive detectors

klendathu2k committed:
Properly handle multiple sensitive detectors mapped to single g2t table.

klendathu2k committed:
Handle BTOF back to (c) 2010.

klendathu2k committed:
Update MTD.  Add BBC, VPD and ETOF.

klendathu2k committed:
Handle inner TPC padrows from before the iTPC, during the single sector 20 test, and the full install.

klendathu2k committed:
Update the volume id factory.

klendathu2k committed:
Allows instrospection about the detector models which are applied in a given geometry tag.

klendathu2k committed:
Remove unused struct.

klendathu2k committed:
Remove actions associated with the ancient (now removed) G3 VMC package.

klendathu2k committed:
Example macro for running geant4 ala embedding.

klendathu2k committed:
Ensure that the geant4 maker is instantiated only once..

...

klendathu2k committed:
Dummy routines should live in the StarGenerator area and be shipped as a 'stubs' library.

klendathu2k committed:
Make configuration available via attributes which, if set, will result in events generated on each Make of the chain.

klendathu2k committed:
Additional verbosity.

klendathu2k committed:
Removed unused TObjArray dependency.  Clingwrap TClonesArray to avoid runtime issues.

klendathu2k committed:
Clingwraps.  Allow configuration purely at the 'maker' level.

klendathu2k committed:
Avoid runtime problems w/ root6

...

klendathu2k committed:
Additional chain options aimed at supporting geant4 embedding.

klendathu2k committed:
Should have been part of 8dc24440.

klendathu2k committed:
Early geometry initialization if geant4star maker exists.

klendathu2k committed:
Define an embedding mode.

1. Geometry option will be looked for in the bfc, physicssim chains.
   Finally lookup via call to GetDatabase.
2. Under embed mode, responsability for setting the vertex will be
   under the control of an embedding maker.

klendathu2k committed:
ignore

klendathu2k committed:
Initial working root6 compatible embedding macro.  Only supports single job for now.

invalid-email-address committed:
Add README.md.  Add documentation block for each class.

klendathu2k committed:
Update documentation.
